janus_server/src/hpke.rs

```diff
@@ -300,35 +428,177 @@ impl HpkeRecipient {
         &self,
         ciphertext: &HpkeCiphertext,
         associated_data: &[u8],
+    ) -> Result<Vec<u8>, Error> {
+        let application_info = HpkeApplicationInfo::new(
+            self.task_id,
+            self.label,
+            self.sender_role,
+            self.recipient_role,
+        );
+        self.open_internal(ciphertext, application_info, associated_data)
+    }
+
+    /// Decrypt `ciphertext` and return the plaintext, but with a directly-specified application
+    /// information byte string. In normal operation, this is called by [open] with the
+    /// PPM-specified domain separation information. Test may use this directly to provide non-PPM
+    /// application information byte strings, for example to check against the HPKE RFC's test
+    /// vectors.
+    fn open_internal(
+        &self,
+        ciphertext: &HpkeCiphertext,
+        application_info: HpkeApplicationInfo,
+        associated_data: &[u8],
     ) -> Result<Vec<u8>, Error> {
         // We must manually dispatch to each possible specialization of open
         match (self.config.kem_id, self.config.kdf_id, self.config.aead_id) {
+            (HpkeKemId::P256HkdfSha256, HpkeKdfId::HkdfSha256, HpkeAeadId::Aes128Gcm) => {
```

Comment:
Same comments as above re: reducing repetition by matching to a function, and matching the ordering of the tuple with the order of the generic type parameters.